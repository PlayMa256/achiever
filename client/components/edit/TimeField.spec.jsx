import React from 'react';
import { shallow } from 'enzyme';

import TimeField from './TimeField';

Math.random = jest.fn(() => '12345');

describe('TimeField', () => {
	describe('render', () => {
		it('should render properly with min props', () => {
			const focusFn = jest.fn();
			const hourTimeField = shallow(<TimeField
				mode="hours"
				onFocus={focusFn}
			/>);
			expect(hourTimeField).toMatchSnapshot();
			expect(focusFn).not.toHaveBeenCalledWith('hours');

			const focusedTimeField = shallow(<TimeField
				shouldHaveFocus
				mode="minutes"
				onFocus={focusFn}
			/>);
			expect(focusedTimeField).toMatchSnapshot();
		});
	});
	describe('_changeTime', () => {
		it('should call onChange and validate it', () => {
			const onChange = jest.fn();
			let wrapper = shallow(<TimeField
				mode="minutes"
				onFocus={jest.fn()}
				onChange={onChange}
			/>);
			const inputField = wrapper.find('input');
			inputField.simulate('change', { target: { value: 35 } });
			expect(onChange).toHaveBeenCalledWith('minutes', 35);

			wrapper = shallow(<TimeField
				mode="hours"
				onFocus={jest.fn()}
				onChange={onChange}
			/>);
			inputField.simulate('change', { target: { value: 'oi' } });
			expect(onChange).toHaveBeenCalledWith('minutes', 0);

			inputField.simulate('change', { target: { value: null } });
			expect(onChange).toHaveBeenCalledWith('minutes', 0);
		});
	});
	describe('onChooseSuggestion', () => {
		it('should change time as user click on Suggestuion Box', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<TimeField
				mode="minutes"
				onFocus={jest.fn()}
				onChange={onChange}
			/>);
			const SuggestionBox = wrapper.find('SuggestionBox');
			SuggestionBox.props().onChoose(15);
			expect(onChange).toHaveBeenCalledWith('minutes', 15);
		});
	});
	describe('onBlur and onFocus', () => {
		it('should be called on blur of input', () => {
			jest.useFakeTimers();
			const wrapper = shallow(<TimeField
				mode="minutes"
				onFocus={jest.fn()}
				onChange={jest.fn()}
			/>);
			wrapper.instance().input = { select: jest.fn() };
			const { onBlur, onFocus } = wrapper.find('input').props();
			onFocus();
			expect(wrapper.instance().input.select).toHaveBeenCalled();
			jest.runAllTimers();
			expect(wrapper.state('isFocused')).toBeTruthy();
			onBlur();
			jest.runAllTimers();
			expect(wrapper.state('isFocused')).toBeFalsy();
		});
	});
	describe('componentDidUpdate', () => {
		// Faz on instance() do wrapper com a variavel input, na variÃ¡vel coloca um objeto...
		it('should have focus if receive props', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<TimeField
				mode="minutes"
				onFocus={jest.fn()}
				onChange={onChange}
				shouldHaveFocus={false}
			/>);
			wrapper.instance().input = { select: jest.fn(), focus: jest.fn() };

			const focusedProps = wrapper.instance().props;
			wrapper.setProps({ ...focusedProps, shouldHaveFocus: true });
			expect(wrapper.instance().input.focus).toHaveBeenCalledWith();
		});
	});
});
