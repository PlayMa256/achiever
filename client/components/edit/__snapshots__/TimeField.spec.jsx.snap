// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeField render should render properly with min props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeField
    disabled={false}
    handleKeyPress={[Function]}
    mode="hours"
    onChange={[Function]}
    onFocus={[MockFunction]}
    referenceHour={9}
    shouldHaveFocus={false}
    tabIndex={0}
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="hours"
          disabled={false}
          id="hours_12345"
          max="23"
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          placeholder="0"
          tabIndex={1}
          type="number"
          value=""
        />,
        <SuggestionBox
          mode="hours"
          onChoose={[Function]}
          referenceHour={9}
          show={false}
        />,
      ],
      "className": "field",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "hours",
          "disabled": false,
          "id": "hours_12345",
          "max": "23",
          "min": 0,
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onKeyUp": [Function],
          "placeholder": "0",
          "tabIndex": 1,
          "type": "number",
          "value": "",
        },
        "ref": [Function],
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "mode": "hours",
          "onChoose": [Function],
          "referenceHour": 9,
          "show": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="hours"
            disabled={false}
            id="hours_12345"
            max="23"
            min={0}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="0"
            tabIndex={1}
            type="number"
            value=""
          />,
          <SuggestionBox
            mode="hours"
            onChoose={[Function]}
            referenceHour={9}
            show={false}
          />,
        ],
        "className": "field",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "hours",
            "disabled": false,
            "id": "hours_12345",
            "max": "23",
            "min": 0,
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "onKeyUp": [Function],
            "placeholder": "0",
            "tabIndex": 1,
            "type": "number",
            "value": "",
          },
          "ref": [Function],
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "mode": "hours",
            "onChoose": [Function],
            "referenceHour": 9,
            "show": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TimeField render should render properly with min props 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeField
    disabled={false}
    handleKeyPress={[Function]}
    mode="minutes"
    onChange={[Function]}
    onFocus={[MockFunction]}
    referenceHour={9}
    shouldHaveFocus={true}
    tabIndex={0}
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="minutes"
          disabled={false}
          id="minutes_12345"
          max="59"
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          placeholder="00"
          tabIndex={1}
          type="number"
          value=""
        />,
        <SuggestionBox
          mode="minutes"
          onChoose={[Function]}
          referenceHour={9}
          show={false}
        />,
      ],
      "className": "field",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "minutes",
          "disabled": false,
          "id": "minutes_12345",
          "max": "59",
          "min": 0,
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onKeyUp": [Function],
          "placeholder": "00",
          "tabIndex": 1,
          "type": "number",
          "value": "",
        },
        "ref": [Function],
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "mode": "minutes",
          "onChoose": [Function],
          "referenceHour": 9,
          "show": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="minutes"
            disabled={false}
            id="minutes_12345"
            max="59"
            min={0}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="00"
            tabIndex={1}
            type="number"
            value=""
          />,
          <SuggestionBox
            mode="minutes"
            onChoose={[Function]}
            referenceHour={9}
            show={false}
          />,
        ],
        "className": "field",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "minutes",
            "disabled": false,
            "id": "minutes_12345",
            "max": "59",
            "min": 0,
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "onKeyUp": [Function],
            "placeholder": "00",
            "tabIndex": 1,
            "type": "number",
            "value": "",
          },
          "ref": [Function],
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "mode": "minutes",
            "onChoose": [Function],
            "referenceHour": 9,
            "show": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
